admiral/pkg/clusters/routingpolicy_handler.go

```diff
@@ -15,51 +16,201 @@ import (
+	return &RoutingPolicyHandler{RemoteRegistry: rr, ClusterID: cId, RoutingPolicyService: rpProcessor}
+}
+
+type RoutingPolicyProcessor interface {
+	ProcessAddOrUpdate(ctx context.Context, eventType admiral.EventType, newRP *v1.RoutingPolicy, oldRP *v1.RoutingPolicy, dependents map[string]string) error
+	ProcessDependency(ctx context.Context, eventType admiral.EventType, dependency *v1.Dependency) error
+	Delete(ctx context.Context, eventType admiral.EventType, routingPolicy *v1.RoutingPolicy) error
+}
+
+type RoutingPolicyService struct {
 	RemoteRegistry *RemoteRegistry
-	ClusterID      string
 }
 
-type routingPolicyCache struct {
-	// map of routing policies key=environment.identity, value: RoutingPolicy object
-	// only one routing policy per identity + env is allowed
-	identityCache map[string]*v1.RoutingPolicy
-	mutex         *sync.Mutex
+func (r *RoutingPolicyService) ProcessAddOrUpdate(ctx context.Context, eventType admiral.EventType, newRP *v1.RoutingPolicy, oldRP *v1.RoutingPolicy, dependents map[string]string) error {
+	var err error
+	for _, remoteController := range r.RemoteRegistry.remoteControllers {
+		if !common.DoRoutingPolicyForCluster(remoteController.ClusterID) {
+			log.Warnf(LogFormat, eventType, "routingpolicy", newRP.Name, remoteController.ClusterID, "RoutingPolicy disabled for cluster")
+			continue
+		}
+		for _, dependent := range dependents {
+			// Check if the dependent exists in this remoteCluster. If so, we create an envoyFilter with dependent identity as workload selector
+			if _, ok := r.RemoteRegistry.AdmiralCache.IdentityClusterCache.Get(dependent).Copy()[remoteController.ClusterID]; ok {
+				_, err1 := createOrUpdateEnvoyFilter(ctx, remoteController, newRP, eventType, dependent, r.RemoteRegistry.AdmiralCache)
+				if err1 != nil {
+					log.Errorf(LogErrFormat, eventType, "routingpolicy", newRP.Name, remoteController.ClusterID, err)
+					err = common.AppendError(err, err1)
+				} else {
+					log.Infof(LogFormat, eventType, "routingpolicy	", newRP.Name, remoteController.ClusterID, "created envoyfilters")
+				}
+			}
+			id := common.GetRoutingPolicyIdentity(newRP)
+			if oldRP != nil {
+				oldEnv := common.GetRoutingPolicyEnv(oldRP)
+				r.RemoteRegistry.AdmiralCache.RoutingPolicyCache.Delete(id, oldEnv, oldRP.Name)
+			}
+			env := common.GetRoutingPolicyEnv(newRP)
+			r.RemoteRegistry.AdmiralCache.RoutingPolicyCache.Put(id, env, newRP.Name, newRP)
+		}
+	}
+	return err
+}
+
+func (r *RoutingPolicyService) ProcessDependency(ctx context.Context, eventType admiral.EventType, dependency *v1.Dependency) error {
+	//ns := getEnvoyFilterNamespace()
+	newDestinations, _ := getDestinationsToBeProcessed(eventType, dependency, r.RemoteRegistry)
+
+	// for each destination, identify the newRP.
+	for _, dependent := range newDestinations {
+		// if the dependent is in the mesh, then get the newRP
+		policies := r.RemoteRegistry.AdmiralCache.RoutingPolicyCache.GetForIdentity(dependent)
```

Comment:
sorry please correct me if I'm wrong, but why are we trying to get routingPolicies of the destination/client?

Reply:
`newDestinations` is the new service that the client onboarded to. So we need to get all routing policies associated with the destination so we can apply it at the client side.